test/e2e/framework/service_util.go

```diff
@@ -806,6 +807,108 @@ func (j *ServiceTestJig) waitForPodsReady(namespace string, pods []string) error
 	return nil
 }
 
+// CheckServiceReachability : checks when serviceType is -
+// ServiceIP: Service should be accessible from within cluster over it's -
+//  1. clusterIP and servicePort
+//  2. serviceName and servicePort
+// NodePort:
+// - Internal clients should be reachable to Service over -
+//   1. ClusterIP and servicePort
+//   2. serviceName and servicePort
+//   3. Node's internalIP and NodePort
+// - External clients should be reachable to service over -
+//   1. Node's publicIP and NodePort
+// ExternalName: FQDN of kubernetes is used as externalName(for air tight platforms). Request to externalNameService MUST resolve to IP.
+func (j *ServiceTestJig) CheckServiceReachability(f *Framework, svc *v1.Service, pod *v1.Pod) error {
+	svcEndpoints, cmds := []string{}, []string{}
+	svcType := svc.Spec.Type
+	clusterIP := svc.Spec.ClusterIP
+	ports := svc.Spec.Ports
+
+	j.SanityCheckService(svc, svcType)
+
+	switch svcType {
+	case v1.ServiceTypeClusterIP:
+
+		// If .spec.clusterIP is set to "" or "None", ClusterIP is not created, so can not test reachability over clusterIP:servicePort
+		isClusterIPV46, err := regexp.MatchString(RegexIPv4+"||"+RegexIPv6, clusterIP)
+		if err != nil {
+			e2elog.Logf("Unable to parse ClusterIP: %s", clusterIP)
+			return err
+		}
+		ep, err := j.Client.CoreV1().Endpoints(f.Namespace.Name).Get(svc.Name, metav1.GetOptions{})
+		if err != nil {
+			e2elog.Logf("Unable to get endpoints for service: %s", svc.Name)
+			return err
+		}
+		// Endpoints must be available for ClusterIP service. (reachable service)
+		if len(ep.Subsets) == 0 {
+			e2elog.Logf("Service endpoints are not available for service: %s", svc.Name)
+			return nil
```

Comment:
Doesn't here need to return error?
line 868 also does the same, I guess here is intentional.
But messages of line 844 and 846 seem to be errors, could you explain?

Reply:
>But messages of line 844 and 846 seem to be errors, could you explain?

There are certain chances that services are created but no backing pods are attached to the service which leads to no endpoints. So we can not test serviceReachability in this scenarios. `ep.Subsets` expresses that atleast 1 ipaddress subset should backing up that service.

Reply:
>Doesn't here need to return error?

It's good to test serviceReachability only if pods are backing the service. But as service creation is independent entity and serviceReachability may vary based on situation, it's ok to not `return error` instead fallback if pods are not attached.

Reply:
Thanks for pointing this. If we are expecting a service reachability, endpoints MUST be available for services except ExternalName service. So Introduced following code replacement.
```
// Wait for endpoints to be created, this may take longer time if service backing pods are taking longer time to run
err = wait.Poll(1*time.Second, ServiceEndpointsTimeout, func() (bool, error) {
    ep, err := j.Client.CoreV1().Endpoints(f.Namespace.Name).Get(svc.Name, metav1.GetOptions{})
    ExpectNoError(err, "Unable to retrieve endpoints for service: %s", svc.Name)
    if len(ep.Subsets) == 0 {
        return false, nil
    }
    return true, nil
})
ExpectNoError(err, "atleast 1 subset of IP address must be available in service endpoint")
```